torch_geometric/nn/models/explainer.py

```diff
@@ -144,3 +148,226 @@ def to_captum(model: torch.nn.Module, mask_type: str = "edge",
             the index specified. (default: :obj:`None`)
     """
     return CaptumModel(model, mask_type, output_idx)
+
+
+class Explainer(torch.nn.Module):
+    r"""An abstract class for implementing Explainer methods for Graph Neural
+    Networks, *e.g.* :class:`~torch_geometric.nn.GNNExplainer and
+    :class:`~torch_geometric.nn.PGExplainer`. It also provides a general
+    visualization methods for graph attributions.
+
+    Args:
+        model (torch.nn.Module): The GNN module to explain.
+        epochs (int, optional): The number of epochs to train.
+            (default: :obj:`None`)
+        lr (float, optional): The learning rate to apply.
+            (default: :obj:`None`)
+        num_hops (int, optional): The number of hops the :obj:`model` is
+            aggregating information from.
+            If set to :obj:`None`, will automatically try to detect this
+            information based on the number of
+            :class:`~torch_geometric.nn.conv.message_passing.MessagePassing`
+            layers inside :obj:`model`. (default: :obj:`None`)
+        return_type (str, optional): Denotes the type of output from
+            :obj:`model`. Valid inputs are :obj:`"log_prob"` (the model
+            returns the logarithm of probabilities), :obj:`"prob"` (the
+            model returns probabilities), :obj:`"raw"` (the model returns raw
+            scores) and :obj:`"regression"` (the model returns scalars).
+            (default: :obj:`"log_prob"`)
+        log (bool, optional): If set to :obj:`False`, will not log any learning
+            progress. (default: :obj:`True`)
+    """
+
+    # TODO: Change docstring
+    def __init__(self, model: torch.nn.Module, lr: Optional[float] = None,
+                 epochs: Optional[int] = None, num_hops: Optional[int] = None,
+                 return_type: str = 'log_prob', log: bool = False):
+        super().__init__()
+        assert return_type in ['log_prob', 'prob', 'raw', 'regression']
+
+        self.model = model
+        self.lr = lr
+        self.epochs = epochs
+        self.num_hops = num_hops
+        self.return_type = return_type
+        self.log = log
+
+    @property
+    def num_hops(self):
```

Comment:
IMO, `num_hops()` should either return `self._num_hops` or compute the number of hops based on the number of message passing layers, right? And the setter simply sets `self._num_hops`.